rasa/shared/nlu/training_data/formats/rasa_yaml.py

```diff
@@ -470,21 +473,71 @@ def process_training_examples_by_key(
         key_examples: Text,
         example_extraction_predicate=lambda x: x,
     ) -> List[OrderedDict]:
-        from ruamel.yaml.scalarstring import LiteralScalarString
-
         result = []
-        for entity_key, examples in training_examples.items():
 
-            converted_examples = [
-                TrainingDataWriter.generate_list_item(
-                    example_extraction_predicate(example).strip(STRIP_SYMBOLS)
-                )
-                for example in examples
-            ]
+        for entity_key, examples in training_examples.items():
+            converted, intent_metadata = RasaYAMLWriter._convert_training_examples(
+                examples, example_extraction_predicate
+            )
 
             next_item = OrderedDict()
             next_item[key_name] = entity_key
-            next_item[key_examples] = LiteralScalarString("".join(converted_examples))
+            if intent_metadata:
+                next_item[KEY_METADATA] = intent_metadata
+
+            render_as_objects = True in [KEY_METADATA in ex for ex in converted]
+            if render_as_objects:
+                rendered = RasaYAMLWriter._render_training_examples_as_objects(
+                    converted
+                )
+            else:
+                rendered = RasaYAMLWriter._render_training_examples_as_text(converted)
+            next_item[key_examples] = rendered
+
             result.append(next_item)
 
         return result
+
+    @staticmethod
+    def _convert_training_examples(
+        training_examples: List[Dict], example_extraction_predicate=lambda x: x
+    ) -> Tuple[List[Dict], Optional[Dict]]:
+        """Returns converted training examples and potential intent metadata."""
+        result = []
+        intent_metadata = None
+
+        for example in training_examples:
+            converted = {KEY_INTENT_TEXT: example_extraction_predicate(example)}
+
+            if isinstance(example, dict) and KEY_METADATA in example:
+                metadata = example[KEY_METADATA]
+
+                if KEY_METADATA_EXAMPLE in metadata:
+                    converted[KEY_METADATA] = metadata[KEY_METADATA_EXAMPLE]
+
+                if intent_metadata is None and KEY_METADATA_INTENT in metadata:
+                    intent_metadata = metadata[KEY_METADATA_INTENT]
+
+            result.append(converted)
+
+        return result, intent_metadata
+
+    @staticmethod
+    def _render_training_examples_as_objects(examples: List[Dict]) -> List[Dict]:
```

Comment:
do you think it would make sense to add a docstring with an example of result? This whole thing was way easier to understand for me after looking at some examples for this in the tests.

Reply:
Good point, does this one explain better what the function is doing?

Reply:
yep 👍🏻